harness/determined/pytorch/_metric_utils.py

```diff
@@ -0,0 +1,163 @@
+    keys = all_metrics[0].keys()
+    metrics_lists = {key: [m[key] for m in all_metrics] for key in keys}
+
+    return metrics_lists, all_num_batches, num_processes
+
+
+def _average_training_metrics(
+    context: det._core.DistributedContext, per_batch_metrics: List[Dict[str, Any]]
+) -> List[Dict[str, Any]]:
+    """Average training metrics across GPUs"""
+    # TODO (liam): decide whether overhead is acceptable to do this by default.
+    # As part of this effort, we should benchmark zmq and torch.distributed communication
+    # primitives to see which is faster.
+    assert context.size > 1, "Can only average training metrics in multi-GPU training."
+    metrics_timeseries = util._list_to_dict(per_batch_metrics)
+
+    # Gather metrics across ranks onto rank 0 slot.
+    # The combined_timeseries is: dict[metric_name] -> 2d-array.
+    # A measurement is accessed via combined_timeseries[metric_name][process_idx][batch_idx].
+    combined_timeseries, combined_num_batches, num_processes = _combine_metrics_across_processes(
+        context, metrics_timeseries, num_batches=len(per_batch_metrics)
+    )
+
+    if context._is_chief:
+        # We can safely cast variables here because this is all happening on the chief, which
+        # is where we gather metrics.
+        combined_timeseries = cast(Dict[str, List[List[Any]]], combined_timeseries)
+        combined_num_batches = cast(List[int], combined_num_batches)
+
+        # If the value for a metric is a single-element array, the averaging process will
+        # change that into just the element. We record what metrics are single-element arrays
+        # so we can wrap them in an array later (for perfect compatibility with non-averaging
+        # codepath).
+        array_metrics = []
+        for metric_name in combined_timeseries.keys():
+            process_batches = combined_timeseries[metric_name]
+            if isinstance(process_batches[0][0], np.ndarray):
+                array_metrics.append(metric_name)
+
+        num_batches = combined_num_batches[0]  # num_batches matches across data parallel ranks.
+        averaged_metrics_timeseries = {}  # type: Dict[str, List]
+
+        for metric_name in combined_timeseries.keys():
+            averaged_metrics_timeseries[metric_name] = []
+            for batch_idx in range(num_batches):
+                batch = [
+                    combined_timeseries[metric_name][process_idx][batch_idx]
+                    for process_idx in range(cast(int, num_processes))
+                ]
+
+                np_batch = np.array(batch)
+                batch_avg = np.mean(np_batch[np_batch != None])  # noqa: E711
+                if metric_name in array_metrics:
+                    batch_avg = np.array(batch_avg)
+                averaged_metrics_timeseries[metric_name].append(batch_avg)
+        per_batch_metrics = util._dict_to_list(averaged_metrics_timeseries)
+    return per_batch_metrics
```

Comment:
The main assumption we are making is that the metrics are scalars or arrays with one element.  I think we technically allow other types (e.g. lists and dictionaries) for which this approach would break.  

If we want to enable averaging of metrics by default, we should modify this to filter out scalars/length one arrays and then raise a warning to say no averaging is performed for lists/dicts if they are detected in metrics.

Reply:
What about the assumption that "the user wants to average all their metrics" instead of e.g. sum all their metrics?

I've basically never liked how we try to reduce user metrics here, but I've never brought it up with pytorch trial for stability reasons.

What if a user provided e.g. a reducer function for each metric they wanted to reduce?  Or is there a way that we could hook up something like the custom reducers API but which applies to batch metrics?

If your opinion is "the way we average metrics is fine" then what you have is fine.  But if you have reservations about it I would suggest perhaps not supporting averaging training metrics in the first version of deepspeed.

Reply:
Separate issue, same line of code: how are we going to deal with the per_batch_metrics that don't get updated on the workers?

I imagine we'll need to solve that at some point, since people are going to want training metrics in callbacks and they will likely want them to be correct on all workers.

Reply:
Okay, I think it makes sense to not have this on by default for DeepSpeedTrial.  

> Related issue: how are we going to deal with the per_batch_metrics that don't get updated on the workers?
I think the desired behavior is to have metrics reduced per batch across ranks and then the user can optionally specify that they want to average metrics across multiple batches.  

This suggests a different custom reducer interface where the user can specify how to reduce across batches in a single train_step and then across time over multiple train_steps.